src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java

```diff
@@ -0,0 +1,112 @@
+     * and before decorations have been added. Can add custom data to the map using
+     * {@link IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     *
+     * @param modifier The function for modifying the {@link net.minecraft.client.renderer.state.MapRenderState} and adding custom render data.
+     */
+    public void registerMapModifier(BiConsumer<MapItemSavedData, MapRenderState> modifier) {
+        RenderStateExtensions.registerMap(modifier);
+    }
+
+    /**
+     * Registers a render state modifier for {@link MapRenderState.MapDecorationRenderState}s which are run after
+     * vanilla map decoration data has been set. Can add custom data to the map using
+     * {@link IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     *
+     * @param mapDecorationTypeKey Key for the registered {@link MapDecorationType}
+     * @param modifier             The function for modifying the {@link MapRenderState.MapDecorationRenderState} and adding custom render data.
+     */
+    public void registerMapDecorationModifier(ResourceKey<MapDecorationType> mapDecorationTypeKey, MapDecorationRenderStateModifier modifier) {
+        RenderStateExtensions.registerMapDecoration(mapDecorationTypeKey, modifier);
+    }
+
+    private static void ensureParametersMatchBounds(TypeToken<?> baseRenderer) {
+        if (baseRenderer.getType() instanceof ParameterizedType parameterizedType) {
+            Class<?> bound = baseRenderer.getRawType();
+            var parameterized = parameterizedType;
+            do {
+                var userArgs = parameterizedType.getActualTypeArguments();
+                var typeArgs = ((Class) parameterizedType.getRawType()).getTypeParameters();
+                for (int i = 0; i < userArgs.length; i++) {
+                    var userType = userArgs[i];
+                    // ignore wildcards. compilation would catch illegal states, so this must not be an important type to capture
+                    if (userType instanceof WildcardType) continue;
+                    var typeBounds = typeArgs[i].getBounds();
+                    // a generic class with multiple bounds, ie AbstractHoglinRenderer<T extends Mob & HoglinBase>
+                    if (userType instanceof TypeVariable<?> typeVariable) {
+                        var userBounds = typeVariable.getBounds();
+                        // check interface bounds
+                        for (int j = 0; j < typeBounds.length; j++) {
+                            if (!((Class) typeBounds[j]).isAssignableFrom((Class<?>) userBounds[j])) {
+                                throw new IllegalArgumentException("%s does not match actual type parameters %s".formatted(userType, parameterized.getRawType()));
+                            }
+                        }
+                    } else {
+                        // check the base class is the same
+                        if (typeBounds.length != 1 || typeBounds[0] != userType) {
```

Comment:
I'd use `.equals` here rather than `==` -- remember the bounds could themselves be parameterized types, not necessarily Classes